trin-execution/src/trie_walker/filter.rs

```diff
@@ -0,0 +1,120 @@
+use alloy::primitives::{B256, U256};
+use rand::{thread_rng, Rng};
+
+use crate::utils::partial_nibble_path_to_right_padded_b256;
+
+#[derive(Debug, Clone)]
+pub struct Filter {
+    start_prefix: U256,
+    end_prefix: U256,
+}
+
+impl Filter {
+    pub fn new_random_filter(slice_count: u16) -> Self {
+        // if slice_count is 0 or 1, we want to include the whole trie
+        if slice_count == 0 || slice_count == 1 {
+            return Self {
+                start_prefix: U256::ZERO,
+                end_prefix: U256::MAX,
+            };
+        }
+
+        let slice_size = U256::MAX / U256::from(slice_count);
+        let random_slice_index = thread_rng().gen_range(0..slice_count);
+
+        let start_prefix = U256::from(random_slice_index) * slice_size;
+        let end_prefix = if random_slice_index == slice_count - 1 {
+            U256::MAX
+        } else {
+            start_prefix + slice_size - U256::from(1)
+        };
+
+        Self {
+            start_prefix,
+            end_prefix,
+        }
+    }
+
+    fn partial_prefix(prefix: U256, shift_amount: usize) -> B256 {
+        prefix
```

Comment:
Any reason not to do one of the following:
1. `prefix | (U256::MAX << shift_amount)` - I think this is the cleanest
    - I would go one step further and say that `shift_amount` is not clear what it means. Instead, I would use opposite value:
    - ```rust
        let prefix_length = 4 * path.len();
        let partial_prefix_mask = !(U256::MAX >> prefix_length);
        let partial_start_prefix = self.start_prefix & partial_prefix_mask;
        let partial_end_prefix = self.end_prefix & partial_prefix_mask;
        ```
2. `(prefix >> shift_amount) << shift_amount` - same idea as yours, but simpler

And in either of those cases, this is simple enough and I would inline it below.

Reply:
I choose 2 as it is a simpler to understand